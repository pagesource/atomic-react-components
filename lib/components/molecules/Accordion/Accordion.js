// @flow
/**
 *
 * Accordion
 *
 */
import React, { useState } from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';
import styles from './Accordion.style';
import type { AccordionHeaderProps, AccordionPanelProps, AccordionProps } from './types';

const AccordionHeader = ({
  children,
  className,
  ariaLabel,
  accordionPanelId,
  isExpanded,
  toggleExpanded,
}: AccordionHeaderProps): Node => {
  const handleClick = () => toggleExpanded(!isExpanded);
  const handleOnKeyPress = (event) => event.keyCode === 13 && toggleExpanded(!isExpanded);

  return (
    <summary
      className={className}
      role="button"
      aria-expanded={isExpanded}
      aria-label={ariaLabel}
      aria-labelledby={accordionPanelId}
      onClick={handleClick}
      onKeyPress={handleOnKeyPress}
      tabIndex="0"
    >
      {children}
    </summary>
  );
};

AccordionHeader.defaultProps = {
  className: 'accordion-header',
  ariaLabel: 'Accordion Header',
};

/**
 * AccordionPanel
 *
 */
const AccordionPanel = ({
  className,
  children,
  id,
  renderPanelBody,
}: AccordionPanelProps): Node => {
  const [isExpanded, toggleExpanded] = useState(false);
  const childrenWithHeaderProps = React.Children.map(children, (accordionHeader) => {
    if (accordionHeader.type.displayName === AccordionHeader.displayName)
      return React.cloneElement(accordionHeader, {
        accordionPanelId: id,
        toggleExpanded,
        isExpanded,
      });
    return false;
  });

  return (
    <details className={classnames('accordion-panel', className)} id={id}>
      {childrenWithHeaderProps}
      {renderPanelBody && (
        <div className={classnames('accordion-panel-content', className)} aria-hidden={!isExpanded}>
          {renderPanelBody()}
        </div>
      )}
    </details>
  );
};

AccordionPanel.defaultProps = {
  className: 'panel',
};

AccordionPanel.header = AccordionHeader;

/**
 * Accordion
 */
const Accordion = ({ className, children }: AccordionProps): Node => (
  <div className={className}>{children}</div>
);

Accordion.defaultProps = {
  className: 'accordion',
};

Accordion.panel = AccordionPanel;

const StyledAccordion: ComponentType<AccordionProps> = styled(Accordion)`
  ${styles};
`;

export default StyledAccordion;

export { Accordion as AccordionVanilla };
