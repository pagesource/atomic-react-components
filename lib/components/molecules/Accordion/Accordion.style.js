import { css } from 'styled-components';

export default css`
  ${(props) => (props.inheritedStyles ? props.inheritedStyles : '')};
  details {
    width: 100%;
    padding: 10px;
    background-color: #eee;
    summary {
      cursor: pointer;
      padding: 20px 0;
    }
    .accordion-panel-content {
      display: none;
      transition: 0.4s;
    }
    &:hover,
    &:active,
    &[open],
    &.active {
      background-color: #ccc;
    }
    &[open],
    &.active {
      .accordion-panel-content {
        display: block;
      }
    }
  }
`;
