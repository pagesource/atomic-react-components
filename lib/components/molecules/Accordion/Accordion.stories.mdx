import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Para from '../../atoms/Para';
import Heading from '../../atoms/Heading';
import Accordion, {AccordionVanilla} from '.';


<Meta title="Molecules/Accordion" component={AccordionVanilla} />

### Accordion
<Preview>
  <Story name="Accordion">
   <Accordion>
      <Accordion.panel
        id="accordion-1"
        renderPanelBody={() => (
          <Para>
            The purpose of lorem ipsum is to create a natural looking block of text (sentence,
            paragraph, page, etc.) that does not distract from the layout. A practice not without
            controversy, laying out pages with meaningless filler text can be very useful when the
            focus is meant to be on design, not content.
          </Para>
        )}
      >
        <Accordion.panel.header>
          <Heading HeadingType="h3" style={{ display: 'inline-block' }}>
            Heading 1
          </Heading>
        </Accordion.panel.header>
      </Accordion.panel>
      <Accordion.panel
        renderPanelBody={() => (
          <Para>
            The purpose of lorem ipsum is to create a natural looking block of text (sentence,
            paragraph, page, etc.) that does not distract from the layout. A practice not without
            controversy, laying out pages with meaningless filler text can be very useful when the
            focus is meant to be on design, not content.
          </Para>
        )}
      >
        <Accordion.panel.header>
          <Heading HeadingType="h3" style={{ display: 'inline-block' }}>
            Heading 2
          </Heading>
        </Accordion.panel.header>
      </Accordion.panel>
    </Accordion>
  </Story>
</Preview>

## Props ðŸ”§

<Props of={AccordionVanilla} />

