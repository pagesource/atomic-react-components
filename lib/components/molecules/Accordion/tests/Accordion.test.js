import React from 'react';
import { mount } from 'enzyme';
import Accordion from '../index';

describe('<Accordion />', () => {
  let AccordionComponent = '';

  beforeEach(() => {
    AccordionComponent = mount(
      <Accordion>
        <Accordion.panel>
          <Accordion.panel.header />
        </Accordion.panel>
      </Accordion>
    );
  });

  test('accordion renders', () => {
    expect(AccordionComponent.exists()).toBe(true);
  });

  test('should render <div/>', () => {
    expect(AccordionComponent.find('div').length).toEqual(1);
  });

  test('should render details', () => {
    expect(AccordionComponent.find('details').length).toEqual(1);
  });

  test('should render summary', () => {
    expect(AccordionComponent.find('summary').length).toEqual(1);
  });

  test('panel opens on click of header', () => {
    AccordionComponent.find('summary').simulate('click');
    expect(AccordionComponent.find('summary').props()).toHaveProperty('aria-expanded', true);
  });

  test('should render correctly', () => {
    expect(AccordionComponent).toMatchSnapshot();
  });
});
