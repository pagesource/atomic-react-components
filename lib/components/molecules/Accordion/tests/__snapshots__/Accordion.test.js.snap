// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> should render correctly 1`] = `
<Accordion__StyledAccordion>
  <Accordion
    className="Accordion__StyledAccordion-vybsad-0 gvcbpS"
  >
    <div
      className="Accordion__StyledAccordion-vybsad-0 gvcbpS"
    >
      <AccordionPanel
        className="panel"
      >
        <details
          className="accordion-panel panel"
        >
          <AccordionHeader
            ariaLabel="Accordion Header"
            className="accordion-header"
            isExpanded={false}
            key=".0"
            toggleExpanded={[Function]}
          >
            <summary
              aria-expanded={false}
              aria-label="Accordion Header"
              className="accordion-header"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            />
          </AccordionHeader>
        </details>
      </AccordionPanel>
    </div>
  </Accordion>
</Accordion__StyledAccordion>
`;
